### PR TITLE
Refactor fixclock hook so doesn't run on LXC

### LINKED ISSUE #1885: Apply better fixclock hook fix for containers
As per #1170, by default TKLBAM will fail on LXC (and Docker and I imagine any other containerisation) - bare metal and "proper" VMs are fine (assuming they can update their own clock - and if not an error is appropriate).

The workaround implemented previously (in buildtasks) simply made the fixclock hook script non-executable at build time. However, it has come to my attention that it can become executable again post initial setup. I assume that happened via updated TKLBAM package.

So a better fix is needed. Use of `systemd-detect-virt` seems like the best approach (i.e. detect containerisation at runtime).

### PR DESCRIPTION
**Pull request by JedMeister**
Refactor fixclock hook so doesn't run on LXC (with appropriate warning message)

Also requires https://github.com/turnkeylinux/buildtasks/pull/78 be merged

Closes https://github.com/turnkeylinux/tracker/issues/1885

### CODE REVIEW

**Review by Christie018 — APPROVED** [review]
Code changes look good to me!

File: contrib/fixclock-hook [hunk 1]
```diff
@@ -26,27 +28,96 @@ You can change the NTP server like this:
...
+    return p.returncode, p.stdout, p.stderr
+
+
+def is_container() -> str:
+    """Check if running in container, if yes, return container type,
+    otherwise 'no'"""
+
+    returncode, stdout, stderr = runit(['systemd-detect-virt', '-c'])
+
+    if stdout.strip() == 'none':
```
Comment (a3s7p) [inline contrib/fixclock-hook:88]: exit code should be checked here instead, and then if it's 0, the container type can be read

> If a virtualization technology is detected, 0 is returned, a non-zero code otherwise.

https://www.freedesktop.org/software/systemd/man/latest/systemd-detect-virt.html

File: contrib/fixclock-hook [hunk 2]
```diff
@@ -26,27 +28,96 @@ You can change the NTP server like this:
...
+        # "proper" VM or bare metal
+        return 'no'
```
Comment (a3s7p) [inline contrib/fixclock-hook:90]: you can just return (and the caller can check against) `None` here if you use `Optional[str]` from `typing` as the return type

File: contrib/fixclock-hook [hunk 3]
```diff
@@ -26,27 +28,96 @@ You can change the NTP server like this:
...
-                                             ERROR=p.stderr)
+                                             ERROR=stderr)
+        echo(msg, stderr=True, fatal=True)
 
-        print(msg, end=' ', file=sys.stderr)
-        sys.exit(1)
 
-def main():
+def dont_fixclock() -> None:
+    retcode, container, stderr = runit(['systemd-detect-virt'])
```
Comment (a3s7p) [inline contrib/fixclock-hook:107]: this call to `systemd-detect-virt` should probably be replaced with passing the return value of `is_container()` to this function

File: contrib/fixclock-hook [hunk 4]
```diff
@@ -26,27 +28,96 @@ You can change the NTP server like this:
...
+
+
+def main() -> None:
     op, state = sys.argv[1:]
 
     if op in ('restore', 'backup') and state == 'pre':
-        fixclock()
+        if is_container() == 'no':
+            fixclock()
+        else:
```
Comment (a3s7p) [inline contrib/fixclock-hook:118]: the container type return value of `is_container()` is never used

File: contrib/fixclock-hook [hunk 5]
```diff
@@ -26,27 +28,96 @@ You can change the NTP server like this:
...
+
+    chmod -x {realpath(sys.argv[0])}
 """
 
-def fixclock():
-    os.environ['PATH'] += ':/usr/sbin'
 
-    command = ["ntpdate", "-u", NTPSERVER]
+def echo(msg: str, stderr: bool = False, fatal: bool = False) -> None:
+    """ Echo messge to terminal (exit 1 if fatal)"""
```
Comment (a3s7p) [inline contrib/fixclock-hook:57]: typo

File: contrib/fixclock-hook [hunk 6]
```diff
@@ -26,27 +28,96 @@ You can change the NTP server like this:
 Unfortunately, something went wrong...
 
 $ERROR
+"""
+
+LXC_TPL = f"""\
```
Comment (a3s7p) [inline contrib/fixclock-hook:33]: should it be called `LXC_TPL` if any container type is detected?

**Review by a3s7p — CHANGES_REQUESTED** [review]
otherwise lgtm

File: contrib/fixclock-hook [hunk 7]
```diff
@@ -26,27 +28,96 @@ You can change the NTP server like this:
...
+##########################
+## FIXCLOCK HOOK FAILED ##
+##########################
+
+Amazon S3 and Duplicity need a UTC synchronized clock, but TKLBAM detected
+that we're running in a container ($CONTAINER).
 
+As the container host manages the local time (and an unprivileged container
```
Comment (a3s7p) [inline contrib/fixclock-hook:41]: if the container has the required privileges, it can change the host clock. i think it is still best to avoid doing that even if we can — principle of least surprise. but then maybe reword the message a bit to clarify that this hook doesn't run inside containers as a matter of principle

File: contrib/fixclock-hook (hunk 2 above)
Comment (a3s7p) [reply contrib/fixclock-hook:90]: https://docs.python.org/3/library/typing.html#typing.Optional

File: contrib/fixclock-hook (hunk 6 above)
Comment (JedMeister) [reply contrib/fixclock-hook:33]: Noticed that myself. I'll change it.

File: contrib/fixclock-hook (hunk 7 above)
Comment (JedMeister) [reply contrib/fixclock-hook:41]: Ok thanks man. I'll tweak the text a little to make it correct.

File: contrib/fixclock-hook (hunk 4 above)
Comment (JedMeister) [reply contrib/fixclock-hook:118]: Yeah good call man. I've refactored things a bit so now there is only one call to `systemd-detect-virt` and the container type is passed to `dont_fixclock()` for inclusion in the message (a nice bonus is that the logic in `main()` now reads better too):
```
-def dont_fixclock() -> None:
-    retcode, container, stderr = runit(['systemd-detect-virt'])
-    msg = Template(LXC_TPL).substitute(CONTAINER=container.strip())
+def dont_fixclock(container: str) -> None:
+    msg = Template(CONT_TPL).substitute(CONTAINER=container)
     echo(msg, stderr=True)
 
 
@@ -113,10 +112,11 @@ def main() -> None:
     op, state = sys.argv[1:]
 
     if op in ('restore', 'backup') and state == 'pre':
-        if is_container() == 'no':
+        container = is_container()
+        if not container:
             fixclock()
         else:
-            dont_fixclock()
+            dont_fixclock(container)
```

**Comment by JedMeister** [comment]
Thanks @a3s7p - Applied your suggested changes, plus also ran shellcheck over the bash bits.

**Comment by JedMeister** [comment]
Argh! I didn't push my last commit (where I actually addressed your points) to this branch before merging...

So I just cherry-picked it into master and pushed: https://github.com/turnkeylinux/tklbam/commit/ed83184c15468faae25a2757074d039de73f135a

**Comment by a3s7p** [comment]
@JedMeister  just noticed another nitpick: https://github.com/turnkeylinux/tklbam/commit/ed83184c15468faae25a2757074d039de73f135a#r136367478

**Comment by JedMeister** [comment]
Thanks @a3s7p - as per direct responses to your comments, I have put the `strip()`s back in.

Whilst not ideal, I again just pushed straight to master: https://github.com/turnkeylinux/tklbam/commit/131a741b0f104227835ef60c4e8739e4add57727
